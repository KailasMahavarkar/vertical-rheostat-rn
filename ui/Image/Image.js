import React, { useState } from 'react';
import {
    TouchableHighlight,
    Image as NativeImage,
    View,
    StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';
import theme from '../theme';
import { NO_CAR_IMAGE } from '../utils/icons';
import { concatString } from '../utils/string';
import { testIdProps } from '../utils/testIdProps';

const propTypes = {
    /** Aspect ratio of image */
    aspectRatio: PropTypes.oneOf(['21:9', '16:9', '4:3', '1:1', '']),
    /** Callback fired when the component is clicked */
    onPress: PropTypes.func,
    /** Alternate Image link if the image doesn't exist */
    placeholderSrc: PropTypes.string,
    /** Default Image if the image takes time to load pass image path in this string using require */
    defaultSrc: PropTypes.any,
    /** Image url */
    src: PropTypes.string,
    /** Value for data testing attribute. */
    'data-testing-id': PropTypes.string.isRequired,
};

const defaultProps = {
    src: '',
    placeholderSrc: NO_CAR_IMAGE,
    aspectRatio: '16:9',
    defaultSrc: require('./assets/no_cars_img.png'),
};

const ASPECT_RATIO = {
    '21:9': 21 / 9,
    '16:9': 16 / 9,
    '4:3': 4 / 3,
    '1:1': 1,
};

const styles = StyleSheet.create({
    container: {
        width: theme.size['full'],
    },
});

const Image = ({
    src,
    placeholderSrc,
    aspectRatio,
    onPress,
    defaultSrc,
    'data-testing-id': testingName,
    ...otherProps
}) => {
    const [hasSrc, setHasSrc] = useState(src.length > 0);

    const handleOnError = () => setHasSrc(false);

    const imageSrc = hasSrc ? src : placeholderSrc;

    const imageStyles = {
        aspectRatio: ASPECT_RATIO[aspectRatio],
    };

    const getImage = () => (
        <NativeImage
            style={imageStyles}
            source={{
                uri: imageSrc,
            }}
            resizeMode="contain"
            onError={handleOnError}
            defaultSource={defaultSrc}
            {...testIdProps(concatString(testingName, 'image'))}
        />
    );

    const getClickableImage = image => (
        <TouchableHighlight onPress={onPress}>{image}</TouchableHighlight>
    );

    const renderImage = !onPress ? getImage() : getClickableImage(getImage());

    return (
        <View style={styles.container} {...otherProps}>
            {renderImage}
        </View>
    );
};

Image.propTypes = propTypes;
Image.defaultProps = defaultProps;

export default Image;
