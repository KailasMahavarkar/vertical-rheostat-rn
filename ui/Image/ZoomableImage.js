import React, { Component } from 'react';
import { View, PanResponder, Image, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import { NO_CAR_IMAGE } from '../utils/icons';
import { concatString } from '../utils/string';
import { testIdProps } from '../utils/testIdProps';

const propTypes = {
    /* Image width */
    width: PropTypes.number.isRequired,
    /** Aspect ratio of image */
    aspectRatio: PropTypes.oneOf(['21:9', '16:9', '4:3', '1:1']),
    /* Value for data-testing-id attribute */
    'data-testing-id': PropTypes.string.isRequired,
    /* Doubletap delay */
    doubleTapDelay: PropTypes.number,
    /* Maximum scale value for zoom feature */
    maxScale: PropTypes.number,
    /** Callback fire when mouse wheel event triggers. */
    onDoubleTap: PropTypes.func,
    /** Callback fire when pinch event triggers. */
    onPinch: PropTypes.func,
    /** Pinch provision enabled if true */
    shouldZoomOnPinch: PropTypes.bool,
    /* Callback fired when touchmove event triggers. */
    onTouchMove: PropTypes.func,
    /** Alternate Image link if the image doesn't exist */
    placeholderSrc: PropTypes.string,
    /** Default image if the image takes time to load */
    defaultSrc: PropTypes.any,
    /** Image url */
    src: PropTypes.string,
    /* Custom styles for image */
    style: PropTypes.oneOfType(PropTypes.object, PropTypes.array),
    /* Threshold scale value for first double tap functionality */
    threshold: PropTypes.number,
    /* Custom styles for image wrapper */
    wrapperStyle: PropTypes.oneOfType(PropTypes.object, PropTypes.array),
    /* Threshold swipe value for swiping functionality */
    thresholdSwipeValue: PropTypes.number,
};

const defaultProps = {
    width: 360,
    aspectRatio: '16:9',
    doubleTapDelay: 500,
    maxScale: 3,
    onDoubleTap: () => {},
    shouldZoomOnPinch: false,
    thresholdSwipeValue: 50,
    onPinch: () => {},
    onTouchMove: () => {},
    onTouchRelease: () => {},
    placeholderSrc: NO_CAR_IMAGE,
    defaultSrc: NO_CAR_IMAGE,
    src: '',
    style: null,
    threshold: 1.5,
    wrapperStyle: null,
};

const ASPECT_RATIO = {
    '21:9': 21 / 9,
    '16:9': 16 / 9,
    '4:3': 4 / 3,
    '1:1': 1,
};

const styles = StyleSheet.create({
    wrapperStyle: {
        overflow: 'hidden',
    },
    imageStyle: {
        position: 'absolute',
    },
});

class ZoomableImage extends Component {
    constructor(props) {
        super(props);

        this.state = {
            scale: 1,
            initialX: null,
            initalY: null,
            top: 0,
            left: 0,
            hasSrc: props.src.length > 0,
            lastTap: 0,
            singlePressTimer: 0,
        };

        this.zoomGlobalStates = {
            isMoving: false,
            isZooming: false,
            initialDistance: null,
            initialZoom: 1,
            initialTopWithoutZoom: 0,
            initialLeftWithoutZoom: 0,
            initialTop: 0,
            initialLeft: 0,
            gestureType: null,
            gestureDirection: null,
        };

        this.height = this.getImageHeight(props.width);
    }

    componentDidUpdate(prevProps, prevState) {
        const { resetScale, width } = this.props;
        if (prevProps.resetScale !== resetScale) {
            this.setState({ scale: 1, top: 0, left: 0 });
        }

        if (prevProps.width !== width) {
            this.height = this.getImageHeight(width);
        }
    }

    getImageHeight = width => {
        const { aspectRatio } = this.props;
        return (1 / ASPECT_RATIO[aspectRatio]) * width;
    };

    maxOffset = (offset, imageDimension) => {
        const { scale } = this.state;
        const max = imageDimension - imageDimension * scale;
        if (max >= 0) {
            return 0;
        }
        return offset < max ? max : offset;
    };

    calcOffsetByZoom = () => {
        const { width } = this.props;
        const { scale } = this.state;
        const height = this.getImageHeight(width);

        const xDiff = width * scale - width;
        const yDiff = height * scale - height;
        return {
            left: -xDiff / 2,
            top: -yDiff / 2,
        };
    };

    handlePinch(x1, y1, x2, y2) {
        const dx = Math.abs(x1 - x2);
        const dy = Math.abs(y1 - y2);
        const distance = Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2));

        const center = {
            x: x1 > x2 ? x1 - (x1 - x2) / 2 : x2 - (x2 - x1) / 2,
            y: y1 > y2 ? y1 - (y1 - y2) / 2 : y2 - (y2 - y1) / 2,
        };

        const { scale, top, left } = this.state;
        const { width, maxScale } = this.props;
        const height = this.height;
        const {
            isZooming,
            initialDistance,
            initialZoom,
            initialTopWithoutZoom,
            initialLeftWithoutZoom,
        } = this.zoomGlobalStates;

        if (!isZooming) {
            const offsetByZoom = this.calcOffsetByZoom();
            this.setState({
                initialX: center.x,
                initialY: center.y,
            });
            this.zoomGlobalStates = {
                ...this.zoomGlobalStates,
                isZooming: true,
                initialDistance: distance,
                initialZoom: scale,
                initialTopWithoutZoom: top - offsetByZoom.top,
                initialLeftWithoutZoom: left - offsetByZoom.left,
            };
        } else {
            const touchZoom = distance / initialDistance;
            const scale =
                touchZoom * initialZoom > 1 ? touchZoom * initialZoom : 1;

            const offsetByZoom = this.calcOffsetByZoom();
            const left = initialLeftWithoutZoom * touchZoom + offsetByZoom.left;
            const top = initialTopWithoutZoom * touchZoom + offsetByZoom.top;

            if (scale < maxScale + 0.1) {
                this.setState({
                    scale,
                    left: left > 0 ? 0 : this.maxOffset(left, width),
                    top: top > 0 ? 0 : this.maxOffset(top, height),
                });
            }
        }
    }

    handleTouchMove(event, x, y, gestureState) {
        const { top, left, initialX, initialY, scale } = this.state;
        const { width, thresholdSwipeValue, onTouchMove } = this.props;
        const height = this.height;
        const { isMoving, initialTop, initialLeft } = this.zoomGlobalStates;
        const { dx } = gestureState;
        if (dx < 0 && -1 * dx > thresholdSwipeValue) {
            this.zoomGlobalStates.gestureDirection = 'left';
        } else if (dx > thresholdSwipeValue) {
            this.zoomGlobalStates.gestureDirection = 'right';
        } else {
            this.zoomGlobalStates.gestureDirection = null;
        }
        onTouchMove(event, scale, this.zoomGlobalStates.gestureDirection);
        if (!isMoving) {
            this.setState({
                initialX: x,
                initialY: y,
            });
            this.zoomGlobalStates = {
                ...this.zoomGlobalStates,
                isMoving: true,
                initialTop: top,
                initialLeft: left,
            };
        } else {
            const left = initialLeft + x - initialX;
            const top = initialTop + y - initialY;

            this.setState({
                left: left > 0 ? 0 : this.maxOffset(left, width),
                top: top > 0 ? 0 : this.maxOffset(top, height),
            });
        }
    }

    handlePanResponderMove = (event, gestureState) => {
        const { onPinch, onTouchMove, shouldZoomOnPinch } = this.props;
        const { scale } = this.state;
        const touches = event.nativeEvent.touches;
        const { isZooming } = this.zoomGlobalStates;
        const { vx, vy } = gestureState;
        if (Math.abs(vx) > 0 || Math.abs(vy) > 0) {
            if (touches.length === 2 && shouldZoomOnPinch) {
                this.zoomGlobalStates.gestureType = 'pinch';
                onPinch(event, scale);
                this.handlePinch(
                    touches[0].pageX,
                    touches[0].pageY,
                    touches[1].pageX,
                    touches[1].pageY
                );
            } else if (touches.length === 1 && !isZooming) {
                this.handleTouchMove(
                    event,
                    touches[0].pageX,
                    touches[0].pageY,
                    gestureState
                );
                this.zoomGlobalStates.gestureType = 'shift';
                onTouchMove(event);
            }
        }
    };

    getNextPositionValues = (nextScale, locationX, locationY) => {
        const { top, left, scale } = this.state;
        const incrementalScalePercentage = (nextScale - scale) / scale;
        const translateY = locationY * incrementalScalePercentage;
        const translateX = locationX * incrementalScalePercentage;
        const nextTop = top - translateY;
        const nextLeft = left - translateX;

        return { nextLeft, nextTop };
    };

    handleDoubleTap = e => {
        const { threshold, maxScale, onDoubleTap } = this.props;
        const { scale } = this.state;
        const { locationY, locationX } = e.nativeEvent;

        if (scale < threshold) {
            onDoubleTap(e, threshold);
            const { nextLeft, nextTop } = this.getNextPositionValues(
                threshold,
                locationX,
                locationY
            );

            this.setState({
                scale: threshold,
                left: nextLeft,
                top: nextTop,
            });
        } else if (scale >= threshold && scale < maxScale) {
            onDoubleTap(e, maxScale);
            const { nextLeft, nextTop } = this.getNextPositionValues(
                maxScale,
                locationX,
                locationY
            );

            this.setState({
                scale: maxScale,
                left: nextLeft,
                top: nextTop,
            });
        } else if (scale >= maxScale) {
            this.setState({ scale: 1, left: 0, top: 0 }, () => {
                onDoubleTap(e, 1);
            });
        }
    };

    handleSingleTap = e => {
        this.cancelSinglePressTimer();
        const { doubleTapDelay, onSingleTap } = this.props;
        const { lastTap } = this.state;
        const now = Date.now();
        if (lastTap && now - lastTap < doubleTapDelay) {
            this.handleDoubleTap(e);
        } else {
            this.setState({ lastTap: now });
            const timeout = setTimeout(() => {
                this.setState({ lastTap: 0 });
                if (onSingleTap) {
                    onSingleTap(e);
                }
            }, doubleTapDelay);
            this.setState({ singlePressTimer: timeout });
        }
    };

    cancelSinglePressTimer = () => {
        const { singlePressTimer } = this.state;
        if (singlePressTimer) {
            clearTimeout(singlePressTimer);
            this.setState({ singlePressTimer: 0 });
        }
    };
    handlePanResponderRelease = event => {
        const { onTouchRelease } = this.props;
        const { scale } = this.state;

        this.zoomGlobalStates.gestureType == 'shift' &&
            onTouchRelease &&
            onTouchRelease(
                event,
                scale,
                this.zoomGlobalStates.gestureDirection
            );
        if (!this.zoomGlobalStates.gestureType) {
            this.handleSingleTap(event);
        }
        this.zoomGlobalStates = {
            ...this.zoomGlobalStates,
            isMoving: false,
            isZooming: false,
            gestureType: null,
        };
    };

    componentWillMount() {
        this._panResponder = PanResponder.create({
            onStartShouldSetPanResponder: () => true,
            onPanResponderMove: this.handlePanResponderMove,
            onPanResponderRelease: this.handlePanResponderRelease,
            onShouldBlockNativeResponder: () => true,
        });
    }

    handleOnError = () => {
        this.setState({ hasSrc: false });
    };

    render() {
        const {
            wrapperStyle,
            style,
            src,
            placeholderSrc,
            width,
            'data-testing-id': testingName,
            defaultSrc,
        } = this.props;
        const { top, hasSrc, left, scale } = this.state;
        const height = this.height;
        const imageSrc = hasSrc ? src : placeholderSrc;

        const imgWrapperStyle = [
            styles.wrapperStyle,
            wrapperStyle,
            {
                width,
                height,
            },
        ];

        const imgStyle = [
            styles.imageStyle,
            style,
            {
                top,
                left,
                width: width * scale,
                height: height * scale,
            },
        ];

        return (
            <View
                style={imgWrapperStyle}
                {...this._panResponder.panHandlers}
                {...testIdProps(
                    concatString(testingName, 'zoomable-image-wrapper')
                )}
            >
                <Image
                    style={imgStyle}
                    source={{ uri: imageSrc }}
                    onError={this.handleOnError}
                    defaultSource={defaultSrc}
                    {...testIdProps(
                        concatString(testingName, 'zoomable-image')
                    )}
                />
            </View>
        );
    }
}

ZoomableImage.propTypes = propTypes;
ZoomableImage.defaultProps = defaultProps;

export default ZoomableImage;
