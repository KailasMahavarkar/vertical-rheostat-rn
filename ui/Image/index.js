import Image from './Image';
import ZoomableImage from './ZoomableImage';

export { ZoomableImage };

export default Image;
