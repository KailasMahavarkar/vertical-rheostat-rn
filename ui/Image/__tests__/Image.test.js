import { shallow } from 'enzyme';
import React from 'react';
import Image from '../index';
import { TouchableHighlight } from 'react-native';
const defaultSrc = require('../assets/no_cars_img.png');
describe('<Image />', function() {
    const src =
        'https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80';
    describe('render', function() {
        it('should render an image', function() {
            const component = mount(<Image src={src} />);

            expect(component).toMatchSnapshot();
        });
    });

    describe('component props', function() {
        describe('should render an image with aspect ratio', function() {
            it('should render an image with 21:9 aspect ratio', function() {
                const component = shallow(
                    <Image src={src} aspectRatio="21:9" />
                );
                expect(component).toMatchSnapshot();
            });

            it('should render an image with 16:9 aspect ratio', function() {
                const component = shallow(
                    <Image src={src} aspectRatio="16:9" />
                );
                expect(component).toMatchSnapshot();
            });

            it('should render an image with 4:3 aspect ratio', function() {
                const component = shallow(
                    <Image src={src} aspectRatio="4:3" />
                );
                expect(component).toMatchSnapshot();
            });

            it('should render an image with 1:1 aspect ratio', function() {
                const component = shallow(
                    <Image src={src} aspectRatio="1:1" />
                );
                expect(component).toMatchSnapshot();
            });
        });

        it('should render an image with placeholder image when image is broken', function() {
            const placeholderSrc =
                'https://imgd.aeplcdn.com/0x0/cw/static/used/no-car-images.svg';
            const component = shallow(
                <Image placeholderSrc={placeholderSrc} />
            );
            const imageComponent = component.find('Image');
            expect(imageComponent).toHaveLength(1);
            expect(imageComponent.props('src').source.uri).toEqual(
                placeholderSrc
            );
        });

        it('should fire onPress event', function() {
            const mockFn = jest.fn();
            const component = mount(<Image src={src} onPress={mockFn} />);
            const container = component.find(TouchableHighlight);
            expect(container).toHaveLength(1);
            container
                .first()
                .props()
                .onPress();

            expect(mockFn).toHaveBeenCalled();
        });

        it('should handle error and render placeholder image', function() {
            const component = mount(<Image src="" />);

            const wrapper = component.find('Image').last();

            wrapper.props().onError();

            expect(component).toMatchSnapshot();
        });
        it('should render placeholder image with defaultSource prop', function() {
            const component = mount(<Image defaultSrc={defaultSrc} />);

            const wrapper = component.find('Image').last();

            wrapper.props().onError();

            expect(component).toMatchSnapshot();
        });

        it('should render Image with data-testing-id', function() {
            const wrapper = mount(<Image src={src} data-testing-id="review" />);

            expect(wrapper).toMatchSnapshot();
        });
    });
});
