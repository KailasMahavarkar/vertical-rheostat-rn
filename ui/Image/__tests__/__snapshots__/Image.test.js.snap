// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> component props should handle error and render placeholder image 1`] = `
<Image
  aspectRatio="16:9"
  defaultSrc=""
  placeholderSrc="https://imgd.aeplcdn.com/0x0/no-image/no-image.png"
  src=""
>
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "width": "100%",
        }
      }
    >
      <Image
        accessibilityLabel={null}
        defaultSource=""
        onError={[Function]}
        resizeMode="contain"
        source={
          {
            "uri": "https://imgd.aeplcdn.com/0x0/no-image/no-image.png",
          }
        }
        style={
          {
            "aspectRatio": 1.7777777777777777,
          }
        }
        testID={null}
      >
        <Image
          accessibilityLabel={null}
          defaultSource=""
          onError={[Function]}
          resizeMode="contain"
          source={
            {
              "uri": "https://imgd.aeplcdn.com/0x0/no-image/no-image.png",
            }
          }
          style={
            {
              "aspectRatio": 1.7777777777777777,
            }
          }
          testID={null}
        />
      </Image>
    </View>
  </View>
</Image>
`;

exports[`<Image /> component props should render Image with data-testing-id 1`] = `
<Image
  aspectRatio="16:9"
  data-testing-id="review"
  defaultSrc=""
  placeholderSrc="https://imgd.aeplcdn.com/0x0/no-image/no-image.png"
  src="https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80"
>
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "width": "100%",
        }
      }
    >
      <Image
        accessibilityLabel="review-image"
        defaultSource=""
        onError={[Function]}
        resizeMode="contain"
        source={
          {
            "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
          }
        }
        style={
          {
            "aspectRatio": 1.7777777777777777,
          }
        }
        testID="review-image"
      >
        <Image
          accessibilityLabel="review-image"
          defaultSource=""
          onError={[Function]}
          resizeMode="contain"
          source={
            {
              "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
            }
          }
          style={
            {
              "aspectRatio": 1.7777777777777777,
            }
          }
          testID="review-image"
        />
      </Image>
    </View>
  </View>
</Image>
`;

exports[`<Image /> component props should render an image with aspect ratio should render an image with 1:1 aspect ratio 1`] = `
<View
  style={
    {
      "width": "100%",
    }
  }
>
  <Image
    accessibilityLabel={null}
    defaultSource=""
    onError={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
      }
    }
    style={
      {
        "aspectRatio": 1,
      }
    }
    testID={null}
  />
</View>
`;

exports[`<Image /> component props should render an image with aspect ratio should render an image with 4:3 aspect ratio 1`] = `
<View
  style={
    {
      "width": "100%",
    }
  }
>
  <Image
    accessibilityLabel={null}
    defaultSource=""
    onError={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
      }
    }
    style={
      {
        "aspectRatio": 1.3333333333333333,
      }
    }
    testID={null}
  />
</View>
`;

exports[`<Image /> component props should render an image with aspect ratio should render an image with 16:9 aspect ratio 1`] = `
<View
  style={
    {
      "width": "100%",
    }
  }
>
  <Image
    accessibilityLabel={null}
    defaultSource=""
    onError={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
      }
    }
    style={
      {
        "aspectRatio": 1.7777777777777777,
      }
    }
    testID={null}
  />
</View>
`;

exports[`<Image /> component props should render an image with aspect ratio should render an image with 21:9 aspect ratio 1`] = `
<View
  style={
    {
      "width": "100%",
    }
  }
>
  <Image
    accessibilityLabel={null}
    defaultSource=""
    onError={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
      }
    }
    style={
      {
        "aspectRatio": 2.3333333333333335,
      }
    }
    testID={null}
  />
</View>
`;

exports[`<Image /> component props should render placeholder image with defaultSource prop 1`] = `
<Image
  aspectRatio="16:9"
  defaultSrc=""
  placeholderSrc="https://imgd.aeplcdn.com/0x0/no-image/no-image.png"
  src=""
>
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "width": "100%",
        }
      }
    >
      <Image
        accessibilityLabel={null}
        defaultSource=""
        onError={[Function]}
        resizeMode="contain"
        source={
          {
            "uri": "https://imgd.aeplcdn.com/0x0/no-image/no-image.png",
          }
        }
        style={
          {
            "aspectRatio": 1.7777777777777777,
          }
        }
        testID={null}
      >
        <Image
          accessibilityLabel={null}
          defaultSource=""
          onError={[Function]}
          resizeMode="contain"
          source={
            {
              "uri": "https://imgd.aeplcdn.com/0x0/no-image/no-image.png",
            }
          }
          style={
            {
              "aspectRatio": 1.7777777777777777,
            }
          }
          testID={null}
        />
      </Image>
    </View>
  </View>
</Image>
`;

exports[`<Image /> render should render an image 1`] = `
<Image
  aspectRatio="16:9"
  defaultSrc=""
  placeholderSrc="https://imgd.aeplcdn.com/0x0/no-image/no-image.png"
  src="https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80"
>
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "width": "100%",
        }
      }
    >
      <Image
        accessibilityLabel={null}
        defaultSource=""
        onError={[Function]}
        resizeMode="contain"
        source={
          {
            "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
          }
        }
        style={
          {
            "aspectRatio": 1.7777777777777777,
          }
        }
        testID={null}
      >
        <Image
          accessibilityLabel={null}
          defaultSource=""
          onError={[Function]}
          resizeMode="contain"
          source={
            {
              "uri": "https://imgd.aeplcdn.com/1056x594/cw/ec/36265/Porsche-Macan-Exterior-161084.jpg?wm=0&q=80",
            }
          }
          style={
            {
              "aspectRatio": 1.7777777777777777,
            }
          }
          testID={null}
        />
      </Image>
    </View>
  </View>
</Image>
`;
